### PR TITLE
Add environments docs

### PR DESCRIPTION
**Pull request by Ekrekr**
~Content seems to fit in well as a separate page, what do you think?~

### CODE REVIEW

File: content/docs/dataform-web/how-to-guides/environments-example.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
+---
```
Comment (lewish) [inline content/docs/dataform-web/how-to-guides/environments-example.md]: IMO, you don't need this page, just make the existing one better?

We have lots of pages where we have a description of the thing then some examples. I think it's simpler if we have all the environments stuff in one page.

File: content/docs/dataform-web/how-to-guides/environments-example.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+- Change to the development branch in the dataform app. Add the following files:
+
+  - One of our template `sqlx` files, or anything that will show up in the run logs.
+
+  - A schedule.
+
+  - An environments file with the contents:
+
+```json
```
Comment (lewish) [inline content/docs/dataform-web/how-to-guides/environments-example.md]: IMO the most common use case for environments, and the example we should show should be using commitSHAs not branches (using branches for releases is a nightmare).

I think there should probably be two examples here:

Example 1 - Creating a staging environment
Example 2 - Promoting releases from staging to production

File: content/docs/dataform-web/how-to-guides/environments-example.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+<div className="bp3-callout bp3-icon-info-sign" markdown="1">
+  Environments require <code>@dataform/core >= 1.4.9</code>.
+</div>
+
+## Introduction
+
+This guide will show you how to configure a project with two environments. One environment will be staging, and the other will be production. Typical of this paradigm, staging will be upstream of production, allowing committed code to be tested before being released.
+
+## Creating a project with environments
```
Comment (lewish) [inline content/docs/dataform-web/how-to-guides/environments-example.md]: Honestly I think this is redundant. You can just link to the other how to guide (setting up a project). Duplicating this here just means more places to have to change it in the future!

Comment (Ekrekr) [reply content/docs/dataform-web/how-to-guides/environments-example.md]: Yes now that you've said it I definitely agree

File: content/docs/dataform-web/how-to-guides/environments-example.md (hunk 1 above)
Comment (Ekrekr) [reply content/docs/dataform-web/how-to-guides/environments-example.md]: Sure, works well

File: content/docs/dataform-web/how-to-guides/environments-example.md (hunk 2 above)
Comment (Ekrekr) [reply content/docs/dataform-web/how-to-guides/environments-example.md]: I've removed a lot of the process description, and added a warning about branches instead.

File: content/docs/dataform-web/how-to-guides/environments.md [hunk 4]
```diff
@@ -84,29 +104,45 @@ Once your team or project outgrows the default environment configuration, a typi
 }
 ```
 
-This configuration defines two environments. The `staging` environment runs the project's schedules at the latest version of the
-project's code (as exists on the `master` branch). When you have confirmed that the code works as expected, you can push those
-schedules to production by merging the `master` branch into the `production` branch.
+To update the version of the project running in `production`, change the value of `commitSha`, and then push that change to your `master` branch.
```
[the remark this answers is not in the document]
Comment (Ekrekr) [reply content/docs/dataform-web/how-to-guides/environments.md]: 👍

File: content/docs/dataform-web/how-to-guides/environments.md [hunk 5]
```diff
@@ -84,29 +104,45 @@ Once your team or project outgrows the default environment configuration, a typi
...
+
+### Configuration via branches
+
+<div className="bp3-callout bp3-icon-info-sign" markdown="1">
+  Branches seen on dataform are not remote git branches, but part of our interal data pipeline system. Because of this, a <code>branch</code> specified in a <code>gitReference</code> won't point a branch on dataform, with the exception of the master branch.
```
[the remark this answers is not in the document]
Comment (Ekrekr) [reply content/docs/dataform-web/how-to-guides/environments.md]: 👍

**Comment by Ekrekr** [comment]
@dwl285 pointed out that overriding the default database is the simplest way to differentiate staging and production data, and as most customers don't use redshift the examples are best using that. However as it's not valid in Redshift, I've added a separate example with the schema config overrides.

Another point he raised was that file specific config overrides are needed in some user cases (link for issue to come)

**Comment by Ekrekr** [comment]
image

File: content/docs/dataform-web/how-to-guides/environments.md [hunk 6]
```diff
@@ -66,47 +88,88 @@ Once your team or project outgrows the default environment configuration, a typi
...
-schedules to production by merging the `master` branch into the `production` branch.
+To update the version of the project running in `production`, change the value of `commitSha`, and then push that change to your `master` branch. On GitHub the commit sha can be found by opening the project page, clicking commits, then copying the desired sha from the presented list of commits.
 
-### Enforcing tighter control on changes to an environment
+### Configuration via branches
 
-In some cases it may be preferred to tightly control changes to an environment. For example, you might want to enforce
-that code is only pushed to a `production` environment after a pull request and/or code review.
+<div className="bp3-callout bp3-icon-info-sign" markdown="1">
+  Branches seen on dataform are not remote git branches, but are only used for local development. Because of this, a <code>branch</code> specified in a <code>gitReference</code> won't point a branch on dataform, with the exception of the master branch.
```
Comment (dwl285) [inline content/docs/dataform-web/how-to-guides/environments.md]: grammer: "won't point a branch"

Comment (dwl285) [reply content/docs/dataform-web/how-to-guides/environments.md]: I don't understand these two sentences?

File: content/docs/dataform-web/how-to-guides/environments.md [hunk 7]
```diff
@@ -66,47 +88,88 @@ Once your team or project outgrows the default environment configuration, a typi
...
 
-This example locks the `production` environment to a specific Git commit. To update the version of the project running in
-`production`, you must change the value of `commitSha`, and then push that change to your `master` branch (which may
-require a pull request, depending on your project's settings). This process has the nice property of requiring that any
-change to `production` is recorded in an audit trail (i.e. your project's Git commits).
+To update the version of the project running in `production`, merge the `master` branch in to the `production` branch.
+
+### Config schema overriding
+
+Redshift and Snowflake do not configure multiple databases like alternative providers. Overriding the schema suffix instead of databases are a good solution to this. For example:
```
Comment (dwl285) [inline content/docs/dataform-web/how-to-guides/environments.md]: Snowflake does. I wouldn't mention the warehouse specifically, I'd just say:
"Another approach to separating `production` and `staging` data is to append a suffix to schemas in the staging environment. For example:"

File: content/docs/dataform-web/how-to-guides/environments.md (hunk 6 above)
Comment (Ekrekr) [reply content/docs/dataform-web/how-to-guides/environments.md]: 👍

File: content/docs/dataform-web/how-to-guides/environments.md (hunk 7 above)
Comment (Ekrekr) [reply content/docs/dataform-web/how-to-guides/environments.md]: 👍

**Pull request merged**